Revised db structure and closed some issues

tungyr-one opened the pull request:
closes #1 
closes #2 
closes #6 
closes #8 
closes #9 
closes #10 
closes #11 
closes #12 
closes #15 
closes #16

Commit:
Revised db structure and closed some issues

coder1coder commented on DocumentApp/ClientApp/src/app/_models/CategoryDto.ts:
```diff
@@ -1,7 +1,7 @@
-import { Doc } from "./Doc";
-import { Subcategory } from "./Subcategory";
 
 export interface CategoryDto {
```
в чем разница между Category и CategoryDto ? Оставить 1 модель

tungyr-one replied:
В модели Category коллекция Children типа Category[], и в нее сохраняются подкатегории, приходящие с бека. В модели CategoryDto коллекция Children типа string[], в нее сохраняются новые имена подкатегорий для сохранения в БД. Я пытался создавать новый массив подкатегорий с именами из формы и присваивать его children родительской категории, но при попытки передачи на бэк получаю ошибку 400. Пока что работаю над поиском решения
![2023-02-28_193440](https://user-images.githubusercontent.com/60428677/221921121-c52224b8-bcf7-472f-a93c-d4be0cce2b9a.png)

coder1coder replied:
Это лишний функционал. Принимаем что за раз можно создать только 1 категорию, без дочерних

tungyr-one replied:
В последнем коммите решил эту проблему, не совпадали типы данных у dto фронта и бэка, отсюда была ошибка парсинга.

tungyr-one replied:
Ok, поправлю

coder1coder commented on DocumentApp/ClientApp/src/app/_models/Category.ts:
```diff
@@ -1,10 +1,10 @@
-   docs: Doc[] | null;
+   parentId: number;
+   children: Category[];
+   docs: Doc[];
```
исключить ссылку на документы

tungyr-one replied:
Исключил везде

coder1coder commented on DocumentApp/ClientApp/src/app/category-edit/category-edit.component.ts:
```diff
@@ -16,72 +14,68 @@ import { Observable, tap } from 'rxjs';
 export class CategoryEditComponent {
   editCategoryForm: FormGroup;
-  id:string|null;
+  id:string;
```
а в какой момент идентификатор стал строкой? на бэке вроде не гуида

tungyr-one replied:
А имеет ли смысл кастить id в number, если в методах он необходим только как string? В контроллер он автоматически парсится как int

coder1coder replied:
По такому принципу всё можно сделать строкой и в http клиент отдавать json строкой

coder1coder replied:
Так что нужно сделать идентификатор типов в соответствии с бэком

tungyr-one replied:
Хорошо, я понял, сделаю

coder1coder commented on DocumentApp/ClientApp/src/app/new-doc/new-doc.component.ts:
```diff
@@ -53,33 +53,30 @@ export class NewDocComponent implements OnInit{
-    })
+    this.categoriesService.getCategories()
+    .pipe(
+      tap(categories => this.Categories = categories),
```
кажется можно объединить

tungyr-one replied:
Объединить pipe и tap?

coder1coder commented on DocumentApp/api/DTOs/CategoryNewUpdateDto.cs:
```diff
@@ -8,6 +8,7 @@ namespace api.DTOs
     public class CategoryNewUpdateDto
```
new или update ?)

tungyr-one replied:
update:)

coder1coder commented on DocumentApp/api/DTOs/CategoryNewUpdateDto.cs:
```diff
@@ -8,6 +8,7 @@ namespace api.DTOs
         public string Name { get; set; }
-        public string[] Subcategories { get; set; }
+        public int? ParentId { get; set; }
+        public string[]? Children { get; set; }
```
дочерние элементы должны быть типа что и родительский

tungyr-one replied:
Сделал

coder1coder commented on DocumentApp/api/Data/DataContext.cs:
```diff
@@ -10,21 +10,7 @@ public DataContext(DbContextOptions options) : base(options)
 
-        protected override void OnModelCreating(ModelBuilder modelBuilder){
-            modelBuilder.Entity<DocDb>(d =>{
-            d.Property(p => p.Id).UseIdentityAlwaysColumn();
```
UseIdentityAlwaysColumn кажется не обязательног, конвенеция наименования EF и так это сделает

tungyr-one replied:
Да, отказался от этого подхода. Моя ошибка была в том, что я явно указывал Id сущностей в seed файле и пытался позже понять почему у меня конфликтуют Id вновь создаваемыx сущностей и существующих в БД

coder1coder commented on DocumentApp/api/Data/Repositories/CategoriesRepository.cs:
```diff
@@ -23,21 +23,20 @@ public CategoriesRepository(DataContext context, IMapper mapper)
       {
             return await _context.Categories
-            .Include(d => d.Subcategories)
+            .Include(d => d.Docs)
```
не не не не, никаких инклюдов документов в категориях

tungyr-one replied:
Ok. Могу я спросить почему?

coder1coder replied:
Потому что метод явно говорит о том что нужно вернуть категории. А что если в одной категории будет 3 млн документов? Сколько времени уйдет на выполнение запроса?

tungyr-one replied:
Понял, спасибо

coder1coder commented on DocumentApp/api/Data/Repositories/CategoriesRepository.cs:
```diff
@@ -23,21 +23,20 @@ public CategoriesRepository(DataContext context, IMapper mapper)
       {
-            return await _context.Categories.Include(c=> c.Subcategories)            
+            return await _context.Categories.Include(c=> c.Children)            
             .FirstOrDefaultAsync(c => c.Id == id);
```
AsNoTracking не помешает везде где есть методы на чтение

tungyr-one replied:
Ok

coder1coder commented on DocumentApp/api/Data/Repositories/DocsRepository.cs:
```diff
@@ -21,8 +21,7 @@ public DocsRepository(DataContext context)
         public async Task<IEnumerable<DocDb>> GetDocsAsync()
         {
```
`
 public Task<DocDb[]> GetDocsAsync()
 => _context.Docs.AsNoTracking()
            .Include(d => d.Category)        
            .ToArrayAsync();
`

tungyr-one replied:
Got it

coder1coder commented on DocumentApp/api/Entities/CategoryDb.cs:
```diff
@@ -9,11 +9,11 @@ namespace api.Entities
     public class CategoryDb
     {
         [System.ComponentModel.DataAnnotations.Key]
```
все правила через FluentApi

tungyr-one replied:
Поубирал все

Commit:
Added app address to config, summaries to code and services to api

Commit:
Amended category edit feature

Commit:
Improved some pull request problems

Commit:
Improved new category transfer to backend

Commit:
Closed bunch of last issues

closes #3
closes #4
closes #5
closes #7
closes #13

Commit:
Fixed problems with bootstrap

Commit:
Changed client id types to number

Commit:
Merge branch categoryEdit into Improvements

Commit:
Simplified category creation feature

Commit:
Moved manage category from Nav menu component to Manage category

coder1coder commented on DocumentApp/ClientApp/src/app/_models/doc.ts:
```diff
@@ -1,12 +1,12 @@
    author: string;
    categoryId: number;
-   subcategoryId: number;
    categoryName: string;
```
кажется categoryName здесь лишнее

tungyr-one replied:
На следующий pr

coder1coder commented on DocumentApp/ClientApp/src/app/_services/category.service.ts:
```diff
@@ -10,39 +10,108 @@ import { Category } from '../_models/Category';
   baseUrl = environment.apiUrl;
   categories: Category[];
+  categoriesNamesWithPrefix: string[] = [];
+  prefix = "-- ";
```
Если префикс нужен для отображения вложенных элементов, то это делается средставами scss/css

tungyr-one replied:
На следующий pr

coder1coder commented on DocumentApp/ClientApp/src/app/category-edit/category-edit.component.ts:
```diff
@@ -16,72 +13,55 @@ import { Observable, tap } from 'rxjs';
-      subcategories: subcategories
-    }
-    console.log('updateCategory:', updateCategory);
+  applyChanges(category:string){
```
пустой метод, выпилить

tungyr-one replied:
Выпилено

coder1coder commented on DocumentApp/api/DTOs/DocNewDto.cs:
```diff
@@ -5,14 +5,16 @@
         public string Author { get; set; }
-        public string CategoryName { get; set; }
-        public string SubcategoryName { get; set; }        
+        public string CategoryName { get; set; }        
```
У категории разве нет идентификатора? Почему мы при создании документа ссылаемся на его наименование а не идентификатор?

tungyr-one replied:
Да, действительно, заменил CategoryName на CategoryId

coder1coder commented on DocumentApp/api/Entities/DocDb.cs:
```diff
@@ -3,19 +3,17 @@
         public int Id { get; set; }
         public string Name { get; set; }             
         public DateTime Created { get; set; }
         public string Version { get; set; }
```
А почему версия представляет из себя строку? Почему не число?

tungyr-one replied:
На следующий pr

coder1coder commented on DocumentApp/api/Interfaces/RepositoriesInterfaces/ICategoriesRepository.cs:
```diff
@@ -0,0 +1,52 @@
+      ///<summary>
+      /// Checks if category exists
+      ///</summary>
+      Task<bool> CategoryExists(string name);
```
Идентифицировать категорию по наименованию плохая затея

tungyr-one replied:
Переделано

coder1coder commented on DocumentApp/api/Program.cs:
```diff
@@ -26,14 +26,13 @@ public static async Task Main(string[] args)
                 var context = services.GetRequiredService<DataContext>();
                 await context.Database.MigrateAsync();
                 await Seed.SeedData(context);
+                await host.RunAsync();
```
Это надо вынести в finally

tungyr-one replied:
Финализировано

coder1coder commented on DocumentApp/api/Services/CategoriesService.cs:
```diff
@@ -0,0 +1,66 @@
+      {
+          _categoriesRepository.Delete(id);
+
+            if (await _categoriesRepository.SaveAllAsync()) return true;
```
можно упростить

tungyr-one replied:
Упрощено в обоих сервисах

coder1coder reviewed (approved):
Обратить внимание на комментарии

Commit:
Replaced CategoryName to CategoryId in DocNewDto and other stuff

Commit:
Improvements for services SaveAllAsync() and other stuff

Simplified services SaveAllAsync() funcs, moved RunAsync to finally, changed string parameter to int type in CategoryExists() func
